### PR TITLE
Freeze docker pods

### LINKED ISSUE #7: Freeze docker containers
The container-freezer should be able to freeze containers running on docker. We can use the [freeze-proxy proof of concept](https://github.com/julz/freeze-proxy/blob/master/pkg/freeze/docker/pause.go) as a guide on how to implement.

### PR DESCRIPTION
**Pull request by psschwei**
# Changes

- Allows the container-freezer service to freeze containers using the docker runtime.

Fixes #7 

Note: there's probably a common CRI package that can be extracted out of the Docker and Containerd ones, but I'd like to just focus here on getting the Docker functionality available, then we can refactor in a follow-up PR to clean everything up.

### CODE REVIEW

**Comment by julz** [comment]
/unassign

please tag me back in when this is ready for a look / non-WIP :)

**Comment by psschwei** [comment]
@julz this should be ready to go now

**Comment by psschwei** [comment]
/assign @julz 

There's some refactoring to be done with the commonalities between the docker and containerd packages, but I'd like to hold off on that for a separate PR and for this one just focus on the capability to pause/resume Docker pods.

**Comment by psschwei** [comment]
Just tossing this out there, but since [docker runtime support is being removed in Kubernetes 1.24](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md#no-really-you-must-read-this-before-you-upgrade), we would only be supporting this until early 2023 (if my back of the napkin calculation for when Knative stops supporting k8s v1.23 is right)...

File: pkg/freeze/docker/pause.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+func NewCRI() (*DockerCRI, error) {
+	c, err := dockerapi.NewClientWithOpts(dockerapi.WithHost(defaultDockerUri),
+		dockerapi.WithVersion(version))
+	if err != nil {
+		return nil, err
+	}
+	return &DockerCRI{client: c}, nil
+}
+
+func (d *Docker) Freeze(ctx context.Context, podUID string) error {
```
Comment (julz) [inline pkg/freeze/docker/pause.go:42]: would it be possible (with some renaming :)) to use one top-level FreezeThawer for docker/crio/containerd and isolate all the differences to the CRI wrappers?

**Comment by julz** [comment]
> Just tossing this out there, but since docker runtime support is being removed in Kubernetes 1.24, we would only be supporting this until early 2023 (if my back of the napkin calculation for when Knative stops supporting k8s v1.23 is right)...

Oops just saw this comment. Yeah, I could see an argument for just doing CRIO and containerd for now, unless someone shouts super loud that they really Really want docker support.

**Comment by psschwei** [comment]
> Oops just saw this comment. Yeah, I could see an argument for just doing CRIO and containerd for now, unless someone shouts super loud that they really Really want docker support.

Ok, going to close this PR for now. Will reopen if we get requests for Docker support.

**Pull request closed without merging**